File: 백준/Gold/7576. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int x, y, cnt;
        public Node(int x, int y, int cnt) {
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        int[][] map = new int[N][M];
        int need = 0;
        Queue<Node> que = new LinkedList<>();
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        for(int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                if(map[i][j] == 0) need++;
                else if(map[i][j] == 1) que.add(new Node(i, j, 0));
            }
        }
        if(need == 0) {
            System.out.println(0);
            return;
        }
        while(!que.isEmpty()) {
            Node node = que.poll();
            for(int i = 0; i < 4; i++) {
                int nx = node.x + dx[i];
                int ny = node.y + dy[i];
                if(nx < 0 || ny < 0 || nx >= N || ny >= M || map[nx][ny] != 0) continue;
                que.add(new Node(nx, ny, node.cnt + 1));
                map[nx][ny] = 1;
                need--;
                if(need == 0) {
                    System.out.println(node.cnt + 1);
                    return;
                }
            }
        }
        System.out.println(-1);
    }
}

File: 백준/Gold/7576. 토마토/토마토.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.head >= this.arr.length;
  }
}

const [[M, N], ...map] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
let sum = M * N;
let tomato = 0;
const que = new Queue();
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
for(let i = 0; i < N; i++) {
  for(let j = 0; j < M; j++) {
    if(map[i][j] === 1) {
      que.add({x : j, y : i, cnt : 0});
      tomato++;
    } else if(map[i][j] === -1) {
      sum--;
    }
  }
}
if(sum === tomato) {
  console.log(0);
  return;
}
let day = 0;
while(!que.isEmpty()) {
  const node = que.poll();
  day = node.cnt;
  for(let i = 0; i < 4; i++) {
    const nx = node.x + dx[i];
    const ny = node.y + dy[i];
    if(nx < 0 || ny < 0 || nx >= M || ny >= N || map[ny][nx] != 0) continue;
    que.add({x : nx, y : ny, cnt : node.cnt + 1});
    map[ny][nx] = 1;
    tomato++;
  }
}
console.log(sum === tomato ? day : -1);